java/ql/src/Telemetry/AutomodelSharedCharacteristics.qll

```diff
@@ -0,0 +1,281 @@
+float maximalConfidence() { result = 1.0 }
+
+float highConfidence() { result = 0.9 }
+
+float mediumConfidence() { result = 0.6 }
+
+/**
+ * A specification of how to  instantiate the shared characteristics for a given candidate class.
+ *
+ * The `CandidateSig` implementation specifies a type to use for Endpoints (eg., `ParameterNode`), as well as a type
+ * to label endpoint classes (the `EndpointType`). One of the endpoint classes needs to be a 'negative' class, meaning
+ *   "not any of the other known endpoint types".
+ */
+signature module CandidateSig {
+  class Endpoint;
```

Comment:
Can we make this signature stricter in what is an acceptable `Endpoint` class? I'm thinking ahead about how to share this between application and framework mode, so we may not want to incorporate all these constraints right away, but I want to start planning for them now. The two I have in mind are:
- `Endpoint` must be a subtype of `DataFlow::Node` (application-mode endpoints are `ExprNode`s, framework-mode endpoints are `ParameterNode`s)
- `Endpoint` must expose a predicate (say `getCallable`) to tell us the corresponding `Callable`. For an `ExprNode` we go arg -> enclosing call -> callable target, for a `ParameterNode` we just use `getEnclosingCallable`. Right now I see we call `e.getEnclosingCallable()` throughout this PR, which makes me think we'll have a lot of refactoring to include application mode.

Reply:
> * Endpoint must be a subtype of DataFlow::Node (application-mode endpoints are ExprNodes, framework-mode endpoints are ParameterNodes)

I'd like to keep the AutomodelSharedCharacteristic file truly _shared_ by not relying on any Java library concepts (like java-dataflow-nodes) in here. I'm following, the way I see it, patterns that are used by libraries in the shared directory, eg: https://github.com/github/codeql/blob/bd3aaf0306d4e0382ae1cda051a63926f0a6dee3/shared/regex/codeql/regex/RegexTreeView.qll#L13.

My plan is to make a AutomodelJavaSharedCharacteristics file (perhaps with a shorter name) when adding application mode, where such a constraint would be added. I could do that now, but maybe that would be a layer of abstraction that's more useful to build when we actually need it. LMK your thoughts!

> * Endpoint must expose a predicate (say getCallable) to tell us the corresponding Callable. For an ExprNode we go arg -> enclosing call -> callable target, for a ParameterNode we just use getEnclosingCallable. Right now I see we call e.getEnclosingCallable() throughout this PR, which makes me think we'll have a lot of refactoring to include application mode.

In application mode, there won't be any callables at all for all arguments to unknown implementations :-) or am I mistaken? Additionally, there won't be much refactoring for application mode, as those implementations are already built in Tiferet's branch. Out of the characteristics in Tiferet's branch, we'll need to copy over all of the characteristics that don't come for free from the shared implementation and add them to application-mode's qll file.

Reply:
> I'd like to keep the AutomodelSharedCharacteristic file truly shared by not relying on any Java library concepts (like java-dataflow-nodes) in here.

Yes that makes sense. I think you can achieve both our constraints by including the `Node` class as part of the signature, or parameterising the signature with the `Node` type.

Reply:
> In application mode, there won't be any callables at all for all arguments to unknown implementations :-) or am I mistaken? 

I thought in application mode, for a `Call` you will still see a corresponding `Callable` in the DB, but `Callable.fromSource()` will not hold for such a `Callable` (since we don't have the source code, only the bytecode signature).

Reply:
> I thought in application mode, for a Call you will still see a corresponding Callable in the DB, but Callable.fromSource() will not hold for such a Callable (since we don't have the source code, only the bytecode signature).

That's interesting.. could be.. I'll test it tomorrow!

Reply:
The principle that a callable always exists, however, is not general — for example: in the javascript implementation, `DataFlow::InvokeNodes` may not have a callee. This is why I've now added the `getCallable` predicate in the Java-specific code, but kept it out of the shared module :+1:

It's been a nice change to add the predicate in the Java-specific code because it factors out most of the mode-specific code.